### PR TITLE
Fixed InsecureUrlResolver to return complete Uri and added a test

### PR DESCRIPTION
**Pull request by chensixx**
Related to #2079 
Fixed InsecureUrlResolver so the url returned will include data store id(s), not just the document id.
Added several tests in insecureUrlResolverTest to see if correct urls can be returned after the change.

### CODE REVIEW

File: .vscode/settings.json
Comment (ChumpChief) [inline .vscode/settings.json:24]: Was this an intentional config change?  I don't know what it does, but if it was unintentional probably revert it.

File: package.json
Comment (ChumpChief) [inline package.json:112]: Similar here, I have no real objection to the formatting changes but it's best to scope out any unintentional changes from your PR.

File: packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts
Comment (ChumpChief) [inline packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts:97]: Use a descriptive variable name (e.g. `pathPart`) rather than just `x`.

File: packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts
Comment (ChumpChief) [inline packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts:65]: This isn't really a full URI, so maybe a more descriptive name (both here and as the helper param) would be something like `documentRelativePath`.

File: packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts
Comment (ChumpChief) [inline packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts:61]: Add a comment here that we're mocking the window since InsecureUrlResolver depends on window.location.

File: packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts
Comment (ChumpChief) [inline packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts:68]: This isn't a response, but rather a url.  Prob name `absoluteUrl`, `expectedAbsoluteUrl`, `expectedResolvedUrl`

File: lerna-package-lock.json
Comment (ChumpChief) [inline lerna-package-lock.json:821]: This doesn't seem bad but I don't really know why these are bumping to 0.1012.2, or why they're currently still at 0.1012.1 when it looks like all our packages already reference ^0.1013.0-0.  @curtisman is this normal?

**Review by ChumpChief — COMMENTED** [review]
Github is showing there are submodule changes but I don't know how to view them.  Do you know what those are and are they expected?

**Comment by agarwal-navin** [comment]
The title to the PR should indicate what your change is about. "Insecure url resolver" doesn't really tell me anything.

Same with the description. Can you please add more detail as to what you are changing and why. If there is anything specific that you want your reviewer to know / look at, you should add that too :-).

File: package.json
Comment (agarwal-navin) [reply package.json]: I agree. If you are already touching this file, then it is okay but otherwise, you should avoid such changes.

---
In reply to: [491242429](https://github.com/microsoft/FluidFramework/pull/3577#discussion_r491242429) [](ancestors = 491242429)

File: packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts
Comment (agarwal-navin) [inline packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts:97]: >const encodedPathArray = pathArray.map((pathPart) => encodeURIComponent(pathPart)); [](start = 12, length = 83)

Do we need to encode the parts of the URL other than the documentId?
When requesting a component from the Container, we do not encode the component Ids. So I would think this is not required.

File: packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts
Comment (agarwal-navin) [inline packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts:102]: >documentUrl = `fluid://${new URL(this.ordererUrl).host}/${encodedTenantId}/${encodedDocId}`; [](start = 12, length = 92)

This is need for both the if-else branches. So, I would rather do something like to avoid code duplication:
let documentUrl = `fluid://${new URL(this.ordererUrl).host}/${encodedTenantId}/${encodedDocId}`;
if (uri === "") {
  // Get encodedUri
  documentUrl += `${encodedUri}`.
}

File: packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts
Comment (agarwal-navin) [inline packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts:91]: >uri: string [](start = 46, length = 11)

This is not the complete uri. But rather the componentPath or you could just use documentRelativePath from above.

File: packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts
Comment (agarwal-navin) [inline packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts:65]: This does not look completely right. There might not be a "/" after the documentId. The url could just be "https://localhost/document1" or something similar. In that case, the documentRelativePath would be incorrect (You should have a test case for this).

I would do something similar to what you did in LocalResolver.
Get the fullPath as follows and send that to the resolveHelper as the uri instead of sending two params:
`const fullPath = parsedUrl.pathname.substr(1);`

In the createNew case above, the fullPath is the "fileName".

You can then get the documentId in resolveHelper from the uri / fullPath simialr to LocalResolver.

This will further simplify the resolveHelper function because you won't need the if-else clause. You can just append the "fullPath" to the final URL:
fluid://${new URL(this.ordererUrl).host}/${encodedTenantId}/${fullPath}

File: packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts
Comment (agarwal-navin) [inline packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts:57]: Consider not using the keyword random but rather some meaningful name. Since these represent dataObjects, maybe dataObject1/dataObject2.

File: packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts
Comment (agarwal-navin) [inline packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts:71]: nit: Use dataObject or dataStore instead of comp because we don't use component anymore. So "dataObjectResolvedUrl" and "dataObjectResponseUrl"

File: packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts
Comment (agarwal-navin) [inline packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts:74]: nit: For code clarity, consider asserting this immediately after getting the resolvedUrl.
Then get the absoluteUrl and assert that it is correct.

Right now, I had to read the code a couple of times to see whats really going on.

File: packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts
Comment (agarwal-navin) [inline packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts:55]: >it("Test RequestUrl for longer url", async () => { [](start = 4, length = 50)

This should say something like "test RequestUrl for url with data object Ids" or something. longer url can just be a url with longer documentId.

File: packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts
Comment (agarwal-navin) [inline packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts:76]: Add a test same as above but with a URL that doesn't have dataObject Id and no "/" at the end of the documentId (To test the case as mentioned here - https://github.com/microsoft/FluidFramework/pull/3577#discussion_r492303056).

File: packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts
Comment (agarwal-navin) [reply packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts:71]: Yeah, testResolvedUrl is fine. Actually, you can use expectedResolvedUrl since this the expected value.

File: packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts
Comment (chensixx) [reply packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts:65]: i don't think just passing one parameter will work with the createNew case. Because I can't use split("/") on the `filename` if it doesn't have any "/". Or I need to add a if-else clause to get documentId.

Comment (chensixx) [reply packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts:65]: also I will have to encode the documentId in fullPath in the final URL

Comment (agarwal-navin) [reply packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts:69]: Okay. Then its fine to pass in the documentId and the part of the url after the document id. However, the following is still incorrect since there may not be a "/" after the documentId:
`const documentRelativePath = parsedUrl.pathname.replace(`/${documentId}/`,"");`

You should remove the trailing "/" after "documentId" in the replace call.

Comment (agarwal-navin) [reply packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts:69]: I tested this and for a URL like "http://localhost/documentUrl", documentRelativePath is evaluated incorrectly.

File: packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts
Comment (agarwal-navin) [inline packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts:72]: >const dataStoreId = "dataStore"; [](start = 8, length = 32)

nit: Move this declaration just before its usage below.

File: packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts
Comment (chensixx) [reply packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts:69]: if i just remove the "/" after documentId in the replace call, i will still have to handle the "/" if there IS something after documentId in the url. So i added a if clause, please let me know if it can be improved.

Comment (agarwal-navin) [reply packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts]: Yeah, if there is something after the "/" or if the url is "http://localhost/documentUrl/", then you need to handle the extra "/". You can easily do that in the "resolveHelper" by not appending a "/" when adding the documentRelativePath:
``` documentUrl += documentRelativePath`; ```
Actually, that made me realize that in resolveHelper you don't need to check if documentRelativePath is empty, you can always add it to the documentUrl. I will leave a comment there for this as well :-).

I think what you have done works but it feels weird. That "if" condition is unnecessary and is slightly wrong (although the final result is right):
If the url is something like "https://localhost/documentUrl/", the "if" condition is false since the parsedUrl will be "documentUrl/" whereas documentId is "documentUrl". So, you end up trying to get the documentRelativePath which does not seem like the right thing to do (even though it works).

Also, I am not a fan of using "replace". I would rather use "slice". So, the code would look something like:
```
const fullPath = parsedUrl.pathname.substr(1);
const documentId = fullPath.split("/")[0];
const documentRelativePath = fullPath.slice(documentId.length + 1);
```

I will leave it up to you to decide which one of these to use :-).

File: packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts
Comment (agarwal-navin) [inline packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts:103]: This if condition is not needed. Adding an empty string won't change the documentUrl so you should always add the documentRelativePath.

**Review by ChumpChief — APPROVED** [review]
Sounds like the remaining build issue just needs a test change, so approving pending that change.

File: packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts
Comment (agarwal-navin) [inline packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts:148]: >`fluid://${new URL(ordererUrl).host}/${tenantId}/${fileName}/!@$123/dataStore!@$`; [](start = 36, length = 82)

This expected URL does not seem right. Why would some  of the special characters be removed? I am guessing its dependent on what this parsing does:
`const parsedUrl = new URL(request.url);`
However, I am not sure what the expected behavior should be. We might have to look into it.

It might also be interesting to see what happens if there are special characters in the "documentId" since it is encoded using "encodeURIComponent".

Comment (agarwal-navin) [reply packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts:148]: Oh, this got changed while I was reviewing it :-).

---
In reply to: [496196065](https://github.com/microsoft/FluidFramework/pull/3577#discussion_r496196065) [](ancestors = 496196065)

**Review by agarwal-navin — APPROVED** [review]
Looks good! Great job on the detailed tests! :-)

**Pull request merged**
